File: observe/src/cz/cas/asu/stelweb/fuky/observe/client/Spectrograph.java
/*
 *
 *   Author: Jan Fuchs <fuky@sunstel.asu.cas.cz>
 *   $Date$
 *   $Rev$
 *
 *   Copyright (C) 2008-2011 Astronomical Institute, Academy Sciences of the Czech Republic, v.v.i.
 *
 *   This file is part of Observe (Observing System for Ondrejov).
 *
 *   Observe is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   Observe is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with Observe.  If not, see <http://www.gnu.org/licenses/>.
 */

package cz.cas.asu.stelweb.fuky.observe.client;

import cz.cas.asu.stelweb.fuky.observe.xml.SpectrographInfoXML;

import java.lang.NumberFormatException;
import java.util.HashMap;
import java.util.Map;
import java.awt.Color;
import javax.swing.JLabel;

public class Spectrograph {

    public enum Element {
        // BEGIN glst
        DICHROIC_MIRROR(0),
        SPECTRAL_FILTER(1),
        COLLIMATOR(2),
        FOCUS_700(3),
        FOCUS_400(4),
        STAR_CALIBRATION(5),
        COUDE_OES(6),
        FLAT(7),
        COMP(8),
        EXP_SHUTTER(9),
        CAM_700_SHUTTER(10),
        CAM_400_SHUTTER(11),
        GRATING(12),
        EXP(13),
        SLITH_CAM(14),
        CORR_700(15),
        CORR_400(16),
        NONE_01(17),
        NONE_02(18),
        NONE_03(19),
        COLLIMATOR_OES(20),
        FOCUS_OES(21),
        EXP_SHUTTER_OES(22),
        EXP_OES(23),
        NONE_04(24),
        IOD_FLASK(25),
        COUDE_SLIT_CAMERA_POWER(26),
        OES_SLIT_CAMERA_POWER(27),
        // END glst
        FOCUS_700_POS(28),
        FOCUS_400_POS(29),
        GRATING_POS(30),
        EXP_COUNT(31),
        EXP_FREQ(32),
        EXP_OES_COUNT(33),
        EXP_OES_FREQ(34),
        FOCUS_OES_POS(35),
        ELEMENT_COUNT(36);

        private int index;

        private Element(int index) {
            this.index = index;
        }

        public int getIndex() {
            return this.index;
        }

        public static int getCount() {
            return Element.ELEMENT_COUNT.getIndex();
        }
    }

    private SpectrographInfoXML spectrographInfoXML;
    private String[] values;
    // TODO: udelat globani barvy
    private Color red_color = new Color(255, 204, 204);
    private Color green_color = new Color(204, 255, 204);
    private Color yellow_color = new Color(255, 255, 51);

    public Spectrograph(SpectrographInfoXML spectrographInfoXML) {
        this.spectrographInfoXML = spectrographInfoXML;
        this.values = new String[Element.getCount()];

        String[] glst = spectrographInfoXML.getGlst().split(" ");

        for (int i = 0; i < glst.length; ++i) {
            if (i == Element.getCount()) {
                break;
            }

            this.values[i] = glst[i];
        }

        this.values[Element.FOCUS_700_POS.getIndex()] = spectrographInfoXML.getSpgp_4();
        this.values[Element.FOCUS_400_POS.getIndex()] = spectrographInfoXML.getSpgp_5();
        this.values[Element.GRATING_POS.getIndex()]   = spectrographInfoXML.getSpgp_13();
        this.values[Element.EXP_COUNT.getIndex()]     = spectrographInfoXML.getSpce_14();
        this.values[Element.EXP_FREQ.getIndex()]      = spectrographInfoXML.getSpfe_14();
        this.values[Element.EXP_OES_COUNT.getIndex()] = spectrographInfoXML.getSpce_24();
        this.values[Element.EXP_OES_FREQ.getIndex()]  = spectrographInfoXML.getSpfe_24();
        this.values[Element.FOCUS_OES_POS.getIndex()] = spectrographInfoXML.getSpgp_22();
    }

    private void setValueColor(Map<String, Object> map, int stop, int moving, int timeout) {
        int value = Integer.valueOf((String)map.get("value"));

        if (value == stop) {
            map.put("value", "stop");
            map.put("color", red_color);
        }
        else if (value == moving) {
            map.put("value", "moving");
            map.put("color", red_color);
        }
        else if (value == timeout) {
            map.put("value", "timeout");
            map.put("color", red_color);
        }
    }

    private String gratingPos2Angle(String value) {
        int pos = Integer.valueOf(value);
        double angle = -0.00487106 * pos + 61.7024;
        double degree;
        double minute;

        degree = Math.floor(angle);
        minute = Math.round((angle - degree) * 60);
        
        if (minute == 60) {
            ++degree;
            minute = 0;
        }

        return String.format("%02.0f:%02.0f", degree, minute);
    }

    public String getValue(Element element) {
        return values[element.getIndex()];   
    }
    
    public void setJLabel(JLabel jLabel, Element element) {
        String value = this.values[element.getIndex()];
        Map<String, Object> map = new HashMap<String, Object>();
        map.put("value", value);
        map.put("color", green_color);

        switch (element) {
            case COLLIMATOR:
            case COLLIMATOR_OES:
                String collimator_states[] = {
                    "stop",
                    "open",
                    "closed",
                    "open left",
                    "open right",
                    "moving",
                    "timeout",
                };

                try {
                    int number = Integer.parseInt(value);

                    if (number != 1) {
                        map.put("color", red_color);
                    }

                    value = String.format("%s - %s", value, collimator_states[number]);
                    map.put("value", value);
                }
                catch (Exception e) {
                    map.put("color", red_color);
                }

                break;

            case DICHROIC_MIRROR:
                setValueColor(map, 0, 5, 6);
                break;

            case SLITH_CAM:
            case SPECTRAL_FILTER:
                setValueColor(map, 0, 6, 7);
                break;

            case FOCUS_700:
            case FOCUS_400:
            case FOCUS_OES:
                if (value.equals("0")) {
                    map.put("value", "stop");
                }
                else {
                    setValueColor(map, -1, 1, -1);
                }
                break;

            case COUDE_OES:
                if (value.equals("1")) {
                    map.put("value", "Coude");
                }
                else if (value.equals("2")) {
                    map.put("value", "OES");
                }
                else {
                    setValueColor(map, 0, 3, 4);
                }
                break;

            case STAR_CALIBRATION:
                if (value.equals("1")) {
                    map.put("value", "Star");
                }
                else if (value.equals("2")) {
                    map.put("value", "Calibration");
                }
                else {
                    setValueColor(map, 0, 3, 4);
                }
                break;

            case EXP_SHUTTER:
            case EXP_SHUTTER_OES:
            case CAM_700_SHUTTER:
            case CAM_400_SHUTTER:
                if (value.equals("1")) {
                    map.put("value", "open");
                }
                else if (value.equals("2")) {
                    map.put("value", "closed");
                }
                else {
                    setValueColor(map, 0, 3, 4);
                }
                break;

            case COMP:
            case FLAT:
                if (value.equals("0")) {
                    map.put("value", "off");
                }
                else {
                    map.put("value", "on");
                    map.put("color", yellow_color);
                }
                break;
                
            case COUDE_SLIT_CAMERA_POWER:
            case OES_SLIT_CAMERA_POWER:
                if (value.equals("0")) {
                    map.put("value", "off");
                    map.put("color", red_color);
                }
                else {
                    map.put("value", "on");
                }
                break;

            case GRATING:
                if (value.equals("0")) {
                    map.put("value", "stop");
                }
                else {
                    setValueColor(map, -1, 1, 2);
                }
                break;

            case EXP:
            case EXP_OES:
                if (value.equals("0")) {
                    map.put("value", "stop");
                    map.put("color", red_color);
                }
                else {
                    map.put("value", "recording");
                    map.put("color", green_color);
                }
                break;


            case CORR_700:
            case CORR_400:
                if (value.equals("0")) {
                    map.put("value", "unknown");
                    map.put("color", red_color);
                }
                else if (value.equals("1")) {
                    map.put("value", "in");
                    map.put("color", green_color);
                }
                else if (value.equals("2")) {
                    map.put("value", "out");
                    map.put("color", green_color);
                }
                break;

            case IOD_FLASK:
                if (value.equals("1")) {
                    map.put("value", "on");
                    map.put("color", yellow_color);
                }
                else if (value.equals("2")) {
                    map.put("value", "off");
                }
                else {
                    setValueColor(map, 0, 3, 4);
                }
                break;

            case GRATING_POS:
                jLabel.setText(gratingPos2Angle((String)map.get("value")));
                jLabel.setToolTipText((String)map.get("value"));
                jLabel.setBackground((Color)map.get("color"));
                return;

            case EXP_OES_COUNT:
            case EXP_COUNT:
                double count = Double.parseDouble(value);
                
                if (count >= 1000000.0) {
                    value = String.format("%.3f Mcounts", count / 1000000.0);
                } else if (count >= 1000.0) {
                    value = String.format("%.3f Kcounts", count / 1000.0);
                } else {
                    value = String.format("%.0f counts", count);
                }

                map.put("value", value);
                break;
                
            case FOCUS_700_POS:
            case FOCUS_400_POS:
            case EXP_FREQ:
            case EXP_OES_FREQ:
            case FOCUS_OES_POS:
            default:
                break;
        }

        jLabel.setText((String)map.get("value"));
        jLabel.setBackground((Color)map.get("color"));
    }
}

File: peso/src/spectrograph.c
/*                                      
 *   Author: Jan Fuchs <fuky@sunstel.asu.cas.cz>
 *   $Date$
 *   $Rev$                                          
 *   $URL$                                          
 *                                                        
 *   Copyright (C) 2011-2013 Astronomical Institute, Academy Sciences of the Czech Republic, v.v.i.
 *                                                                                            
 *   This file is part of PESO.
 *                                                                                            
 *   PESO is free software: you can redistribute it and/or modify                          
 *   it under the terms of the GNU General Public License as published by                     
 *   the Free Software Foundation, either version 3 of the License, or                        
 *   (at your option) any later version.                                                      
 *                                                                                            
 *   Observe is distributed in the hope that it will be useful,                               
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of                           
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                            
 *   GNU General Public License for more details.                                             
 *                                                                                            
 *   You should have received a copy of the GNU General Public License                        
 *   along with Observe.  If not, see <http://www.gnu.org/licenses/>.                         
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <time.h>
#include <xmlrpc-c/base.h>
#include <xmlrpc-c/client.h>

#include "bxmlrpc.h"
#include "spectrograph.h"

static xmlrpc_env sgh_env;
static xmlrpc_client *p_sgh_client = NULL;
static xmlrpc_server_info *p_sgh_server_info = NULL;
static char sgh_err_msg[SGH_ERR_MSG_MAX + 1];

static int sgh_is_fault_occurred(xmlrpc_env *p_sgh_env)
{
    if (p_sgh_env->fault_occurred)
    {
        snprintf(sgh_err_msg, SGH_ERR_MSG_MAX, "XML-RPC Fault: %s (%d)",
                p_sgh_env->fault_string, p_sgh_env->fault_code);
        fprintf(stderr, "XML-RPC Fault: %s (%d)\n", p_sgh_env->fault_string,
                p_sgh_env->fault_code);
        return 1;
    }
    else
    {
        return 0;
    }
}

//static int sgh_struct_read(xmlrpc_value *p_result, const char *p_name, char *p_output)
//{
//    xmlrpc_value *p_value;
//    const char *p_str;
//
//    xmlrpc_struct_read_value(&sgh_env, p_result, p_name, &p_value);
//
//    if (sgh_is_fault_occurred(&sgh_env)) {
//        return -1;
//    }
//
//    xmlrpc_read_string(&sgh_env, p_value, &p_str);
//
//    if (sgh_is_fault_occurred(&sgh_env)) {
//        return -1;
//    }
//
//    strncpy(p_output, p_str, SGH_ANSWER_MAX);
//
//    xmlrpc_DECREF(p_value);
//    free((char*)p_str);
//
//    return 0;
//}

static char *sgh_get_item(xmlrpc_value *p_result, const char *p_name)
{
    xmlrpc_value *p_value;
    char *p_str;

    xmlrpc_struct_read_value(&sgh_env, p_result, p_name, &p_value);

    if (sgh_is_fault_occurred(&sgh_env))
    {
        return NULL;
    }

    xmlrpc_read_string(&sgh_env, p_value, (const char **) &p_str);

    if (sgh_is_fault_occurred(&sgh_env))
    {
        return NULL;
    }

    xmlrpc_DECREF(p_value);

    return p_str;
}

static void sgh_free_item(char *p_item)
{
    // TODO: osetrit p_item == NULL
    free((char*) p_item);
}

const char *sgh_get_err_msg(void)
{
    return sgh_err_msg;
}

// (-205.294 * <stupne.desetiny...> + 12667.1)  ....[pocet inkrementu]
int sgh_gratang2gratpos(double gratang)
{
    return (-205.294 * gratang + 12667.1);
}

// (-0.00487106 * <inkrementy> + 61.7024)  .... [uhel: stupne.desetiny...]
double sgh_gratpos2gratang(int gratpos)
{
    return (-0.00487106 * gratpos + 61.7024);
}

void sgh_gratpos2gratang_str(int gratpos, char *p_gratang, int gratang_len)
{
    double gratang;
    double degree;
    double minute;

    gratang = sgh_gratpos2gratang(gratpos);
    degree = floor(gratang);
    minute = round((gratang - degree) * 60);

    if (minute == 60) {
        ++degree;
        minute = 0;
    }

    snprintf(p_gratang, gratang_len, "%.0f:%.0f", degree, minute);
}

int sgh_init(void)
{
    char server_url[SGH_SERVER_URL_MAX + 1];
    char *p_spectrograph_host = getenv("SPECTROGRAPH_HOST");
    char *p_spectrograph_port = getenv("SPECTROGRAPH_PORT");

    bzero(sgh_err_msg, SGH_ERR_MSG_MAX + 1);

    if ((p_spectrograph_host == NULL) || (p_spectrograph_port == NULL))
    {
        snprintf(sgh_err_msg, SGH_ERR_MSG_MAX,
                "Undefined variable: SPECTROGRAPH_HOST=%s, SPECTROGRAPH_PORT=%s",
                p_spectrograph_host, p_spectrograph_port);
        return -1;
    }

    xmlrpc_env_init(&sgh_env);

    if (bxr_client_init() == -1)
    {
        return -1;
    }

    xmlrpc_client_create(&sgh_env, 0, "peso", SVN_REV, NULL, 0, &p_sgh_client);
    if (sgh_is_fault_occurred(&sgh_env))
    {
        fprintf(stderr, "\n%s\n", sgh_err_msg);
        return -1;
    }

    snprintf(server_url, SGH_SERVER_URL_MAX, "http://%s:%s/RPC2",
            p_spectrograph_host, p_spectrograph_port);
    p_sgh_server_info = xmlrpc_server_info_new(&sgh_env, server_url);
    if (sgh_is_fault_occurred(&sgh_env))
    {
        fprintf(stderr, "\n%s\n", sgh_err_msg);
        return -1;
    }

    return 0;
}

int sgh_uninit(void)
{
    if (p_sgh_server_info != NULL)
    {
        xmlrpc_server_info_free(p_sgh_server_info);
        p_sgh_server_info = NULL;
    }

    if (p_sgh_client != NULL)
    {
        xmlrpc_client_destroy(p_sgh_client);
        p_sgh_client = NULL;
    }

    xmlrpc_env_clean(&sgh_env);
    bxr_client_cleanup();

    return 1;
}

// TODO: uvolnovani pameti i pri predcasnem ukonceni
int sgh_spectrograph_execute(char *p_command, char *p_answer)
{
    xmlrpc_value *p_result;
    xmlrpc_value *p_param_array;
    xmlrpc_value *p_item;
    const char *p_str;

    p_param_array = xmlrpc_array_new(&sgh_env);
    p_item = xmlrpc_string_new(&sgh_env, p_command);
    if (sgh_is_fault_occurred(&sgh_env))
    {
        return -1;
    }

    xmlrpc_array_append_item(&sgh_env, p_param_array, p_item);
    if (sgh_is_fault_occurred(&sgh_env))
    {
        return -1;
    }

    xmlrpc_client_call2(&sgh_env, p_sgh_client, p_sgh_server_info,
            "spectrograph_execute", p_param_array, &p_result);
    if (sgh_is_fault_occurred(&sgh_env))
    {
        return -1;
    }

    xmlrpc_read_string(&sgh_env, p_result, &p_str);

    if (sgh_is_fault_occurred(&sgh_env))
    {
        return -1;
    }

    xmlrpc_DECREF(p_result);

    strncpy(p_answer, p_str, SGH_ANSWER_MAX);
    free((char*) p_str);

    xmlrpc_DECREF(p_item);
    xmlrpc_DECREF(p_param_array);

    return 0;
}

int sgh_spectrograph_info(SGH_INFO_T *p_sgh_info)
{
    int i;
    int star_calib;
    char *p_glst;
    char *p_item;
    char *p_save = NULL;
    xmlrpc_value *p_result;
    xmlrpc_value *p_param_array;

    memset(p_sgh_info, 0, sizeof(SGH_INFO_T));

    p_param_array = xmlrpc_array_new(&sgh_env);
    xmlrpc_client_call2(&sgh_env, p_sgh_client, p_sgh_server_info,
            "spectrograph_info", p_param_array, &p_result);
    if (sgh_is_fault_occurred(&sgh_env))
    {
        return -1;
    }

    p_sgh_info->flat = 0;
    p_sgh_info->comp = 0;

    if ((p_glst = sgh_get_item(p_result, "GLST")) == NULL)
    {
        return -1;
    }

    if ((p_item = strtok_r(p_glst, " ", &p_save)) != NULL)
    {
        p_sgh_info->dichroic_mirror = atoi(p_item);
    }

    i = 2;
    while ((p_item = strtok_r(NULL, " ", &p_save)) != NULL)
    {
        switch (i)
        {
        // spectral filter
        case 2:
            p_sgh_info->spectral_filter = atoi(p_item);
            break;

            // collimator
        case 3:
            p_sgh_info->collimator = atoi(p_item);
            break;

            // star/calibration
        case 6:
            star_calib = atoi(p_item);
            p_sgh_info->star_calib = star_calib;

            // 0 = unknown
            // 1 = star
            // 2 = calibration
            // 3 = moving
            // 4 = timeout
            if (star_calib == 2)
            {
                p_sgh_info->flat += 2;
                p_sgh_info->comp += 2;
            }
            break;

            // coude/oes
        case 7:
            p_sgh_info->coude_oes = atoi(p_item);
            break;

            // flat
        case 8:
            // 0 = off
            // 1 = on
            p_sgh_info->flat += atoi(p_item);
            break;

            // comp
        case 9:
            // 0 = off
            // 1 = on
            p_sgh_info->comp += atoi(p_item);
            break;

        // Correction plate 700
        case 16:
            p_sgh_info->correction_plate_700 = atoi(p_item);
            break;

        // Correction plate 400
        case 17:
            p_sgh_info->correction_plate_400 = atoi(p_item);
            break;

            // OES collimator
        case 21:
            p_sgh_info->oes_collimator = atoi(p_item);
            break;

        // OES Iodine cell
        case 26:
            p_sgh_info->oes_iodine_cell = atoi(p_item);
            break;

        default:
            break;
        }

        ++i;
    }

    sgh_free_item(p_glst);

    if ((p_item = sgh_get_item(p_result, "SPFE_14")) != NULL)
    {
        p_sgh_info->exp_freq = atoi(p_item);
        sgh_free_item(p_item);
    }

    if ((p_item = sgh_get_item(p_result, "SPCE_14")) != NULL)
    {
        p_sgh_info->exp_sum = atoi(p_item);
        sgh_free_item(p_item);
    }

    if ((p_item = sgh_get_item(p_result, "SPGP_13")) != NULL)
    {
        p_sgh_info->grating_position = atoi(p_item);
        sgh_free_item(p_item);
    }

    if ((p_item = sgh_get_item(p_result, "SPGP_5")) != NULL)
    {
        p_sgh_info->focus_1400 = atoi(p_item);
        sgh_free_item(p_item);
    }

    if ((p_item = sgh_get_item(p_result, "SPGP_4")) != NULL)
    {
        p_sgh_info->focus_700 = atoi(p_item);
        sgh_free_item(p_item);
    }

    if ((p_item = sgh_get_item(p_result, "SPGP_22")) != NULL)
    {
        p_sgh_info->focus_oes = atoi(p_item);
        sgh_free_item(p_item);
    }

    if ((p_item = sgh_get_item(p_result, "SPGS_19")) != NULL)
    {
        p_sgh_info->coude_temp = atoi(p_item);
        sgh_free_item(p_item);
    }

    if ((p_item = sgh_get_item(p_result, "SPGS_20")) != NULL)
    {
        p_sgh_info->oes_temp = atoi(p_item);
        sgh_free_item(p_item);
    }

    xmlrpc_DECREF(p_result);
    xmlrpc_DECREF(p_param_array);

    return 0;
}

float sgh_temp2human(int temp_raw) {
    float temp;

    temp = temp_raw / (27648/80.) - 30;

    return temp;
}
